### ISSUE TITLE
[suggestion] P2P module Enhancement

### ISSUE THREAD

**Issue by dusmart** [neo-project/neo#2862]
**Summary or problem description**
If bad guys sent lots of valid transactions on chain in priority-ascending order with more than 1000 transactions per second, the consensus can then hardly be reached for a very long time. 
If the attack details are required, please contact @superboyiii. He helped us for identifying this problem. We've tried our best to solve this while we did't get a good solution. Therefore, we issued a new enhancement request here to ask for core developer's solutions.

* consensus module enhancement needed: `TxNotFound` will be a frequent reason for view changes, we'd better solve this issue first
* p2p module enhancement needed: after the flood attack start, the p2p messages whether its high-priority or not, such as ChangeView, can not be sent to other nodes in time, therefore the consensus can not be reached

**Do you have any solution you want to propose?**

* pack the transaction bodies in the PREPARE consensus message
* limit p2p peers' in-flow throughput by some throttle methods

We have discussed the proposal with Roman Khimov privately and he didn't like these two solutions. Roman had other methods for mitigation.

**Neo Version**
- Neo 3

**Where in the software does this update applies to?**
- Consensus
- P2P (TCP)

**Comment by Liaojinghui** [comment]
I have a general question here, how long exactly does it takes to sync 1000 txs across the entire neo network? How about 10,000 txs? what if attackers send each consensus node 1000 txs different valid txs, with different accounts from different IP addresses?

 Before we get another sound solution, i vote for pack transaction bodies in the consensus message.

**Comment by dusmart** [comment]
> I have a general question here, how long exactly does it takes to sync 1000 txs across the entire neo network? How about 10,000 txs? what if attackers send each consensus node 1000 txs different valid txs, with different accounts from different IP addresses?
> 
> Before we get another sound solution, i vote for packing transaction bodies in the consensus message.

Sorry that I wrote a wrong number here. The attack speed should be 10k tx/s. If it's 5k tx/s, the network will work just fine. If the speed is much more larger, then the network will be worse. I'll send the result to you privately. Anyway, this problem should attract our notation.

**Comment by roman-khimov** [comment]
A bit more open discussion really can be helpful in this case, I'll reiterate on some points discussed previously.

First, I don't think you can avoid having an inconsistent (unsynchronized wrt mempool contents) network. Any throttling scheme can be subverted with enough senders (as in `tx.signers[0]`) and nodes (as in machines sending data to the network). So we have to deal with this inconsistency in one way or another.

Sending transaction contents with PrepareRequest is a solution, what I don't like about it is the associated cost it has. Correct me if I'm wrong, but I think dBFT tried to avoid doing exactly that since its inception. Testnet has a MaxBlockSize of 2M, most of this volume will be transactions, so if we're to try sending this large PrepareRequest it will take like 2s for every hop assuming 10Mbit/s. Of course you can have somewhat better connection, but you can also consider additional network stress from the transaction flow, either way I'd say this volume will be noticeable. Even in ideal case (synchronized mempools!) it'll cause some measurable block delay.

So if we're to reject the idea of including transactions into PrepareRequest (but hey, convince me it's fine!), we have two basic routes: either relying on CVs to handle the situation (they're a part of the normal protocol anyway) and trying to make the next proposal more suitable for everyone, or trying to avoid them.

In that sense #2058 (implemented in NeoGo) which tries to optimize the decision by using local data only can be extended to try taking into account data from other nodes in a way described in https://github.com/neo-project/neo/issues/2058#issuecomment-725634596 (hi, @vncoelho!). All nodes requesting CV send a reason for that. And if it's TxNotFound then they at least have got a proposal and they really can provide a bitmask of tx presence for that proposal.

The problem is that if the network is being stressed in a way that causes complete mempool replacements in time comparable with the regular block time this may lead to the transaction data being outdated by the time CV is to happen. That is the new primary may have _none_ of the old transactions in its mempool when it's to make a proposal. This, btw, may defeat the original assumption of #2058 as well, rendering it ineffective.

We can try solving this problem before CV happens though, this requires getting transactions in the same round in some more reliable manner (but avoiding a full copy in PrepareRequest). CNs missing transactions from proposal do perform requests for them from the nearest neighbors which in many cases lead to successful block acceptance. But in case it doesn't we may try sending a new broadcasted RecoverTx message to the primary _before_ CV timer expires. It can either be parameter-less or contain the same bitmask as mentioned above (so it'll be small).

This should be a high-priority message (so that it'd be more likely to propagate in time) that is processed by the primary only. Primary can wait for an f+1 set of these messages and then either broadcast a full set of transactions in a new message or just those that are missed on other CNs.

There can also be some heuristic like not sending RecoverTx if we're missing less than 20% of transactions or if the proposal contains less than 20% of MaxTransactionsPerBlock (numbers are pretty arbitrary here, this needs to be tested of course). RecoverTx may be sent after initial attempt to get transactions from neighbors like we send requests and wait 1/3 of the CV timer and if transactions are not there then we send RecoverTx and wait for the next 2/3 of the same timer.

This may keep the performance we have in the optimistic case (highly synchronized network) while allowing quicker recovery in case there is a problem. Even if we're to have some problems collecting transactions at view 0, the next one will have extended timers and thus higher chance to succeed.

**Comment by vncoelho** [comment]
Thanks for pinging me here.
This is also a important discussion for the current PR #2818.

**Comment by vncoelho** [comment]
> This should be a high-priority message (so that it'd be more likely to propagate in time) that is processed by the primary only. Primary can wait for an f+1 set of these messages and then either broadcast a full set of transactions in a new message or just those that are missed on other CNs.

This may be a good strategy for this edge cases, @roman-khimov.

We can also track these messages on the backups and extend it in a similar way we do with `ExtendByFactor`.

**Comment by vncoelho** [comment]
Regarding my worry on the Conflicts Attribute PR,in fact that can happen nowadays with the ascending fees, as you highlighted: "causing complete mempool replacements".

However, differently, the attacker will eventually need to pay for those `network fees` and consequent `system fees`.

**Comment by Liaojinghui** [comment]
As @roman-khimov has stated, the packing transaction body would cause delayed consensus message processing. Thus, I propose a new solution below that does not require any change to the consensus:

The main reason for the attack is that attackers refresh CNs' mempools with ascending transactions after the primary issues a `prepare message`, causing the CNs to be busy requesting transactions (in the meanwhile, attackers keep sending new transactions to update CNs' mempools to slowdown CNs and cause network congestion).

My new solution would target the `mempool refresh` feature of this attack:

1. If full, mempool freezes itself and stops accepting new transactions, even if with higher transaction fees.
2. Only update mempool with transactions in the `prepare message`, here we can adopt the method proposed by @roman-khimov above.
3. for emergency transactions that need to be processed early, it needs to pay at least 10 times more fees than the highest fee in the last block, 10 GSA if the last block is empty.

focusing on the mempool would pose a minimal impact on existing neo protocols.

**Comment by EdgeDLT** [comment]
> We can try solving this problem before CV happens though, this requires getting transactions in the same round in some more reliable manner (but avoiding a full copy in PrepareRequest).

Intuitively I agree, but it strikes me as a problem that will remain in some form as long as transaction selection remains within the purview of CNs. As @roman-khimov points out, unsynchronized mempools can't really be avoided.

I'd like to propose that we give some consideration to what is known in the Ethereum ecosystem as **proposer-builder separation**. Atomic transaction batches (or even full block bodies) can be constructed by trusted (e.g. non-CN committee members) or even untrusted nodes (à la dBFT 3.0). These can be validated and relayed to (or possibly by) CNs, who simply select the most profitable batches to use in proposed blocks.

I'm skirting over a lot of details, and a lot of the research on this topic is designed with Ethereum in mind, so would need a number of adjustments to work in our case. But there are various potential implications of this architectural adjustment that I think at least merits further investigation. To name a few:

* Partially mitigate the issue of mempool flooding and improve scalability by optimizing resource use on the network (avoid more CVs).
* Enable outside node incentivization (fees can be shared with bundlers/builders, i.e. non-CNs). This in turn can further mitigate the flood risk -- if you have good reason to relay thousands of valid transactions per second, why not become a bundler and profit from them instead?
* MEV and censorship resistance (transactions can be privately shared with bundlers to avoid the traditional public mempool route). Not a silver bullet solution, but can reduce the attack surface considerably.
* A step forward in towards the possibility of having stateless validators, which would be especially beneficial in the long-term. 
* Solve the committee free-rider problem (we could require non-CN committee nodes to be builders/relayers and use it to confirm they are doing work). Currently, a committee member can be a light wallet, [as I'm doing on TestNet](https://testmagnet.explorer.onegate.space/accountprofile/0xe827d76b1629139bacc877169c714a5e6d41ffdd). This threatens consensus stability and network throughput via CVs.

There's more, and I don't claim to be an expert, but I think there are some interesting potentials for us here.

**Comment by vncoelho** [comment]
@Liaojinghui

1. I believe that stop accepting transactions is critical because we have a logic on the P2P for requesting and receiving them, that would block the tx to be received later or we would need to do not add them on the cache.
Perhaps, create another hash list to request them later.
2. This maybe bootleneck the process and what will the nodes do while not on the prepare message, reject all txs because of space being full?
3. That possibility of emergency transactions is great in several ways! I believe we can have a special space and list just for such kind of transactions. They are important even for updating committee and blockchain on-chain rules.

**Comment by vncoelho** [comment]
> I'd like to propose that we give some consideration to what is known in the Ethereum ecosystem as proposer-builder separation. Atomic transaction batches (or even full block bodies) can be constructed by trusted (e.g. non-CN committee members) or even untrusted nodes (à la dBFT 3.0). These can be validated and relayed to (or possibly by) CNs, who simply select the most profitable batches to use in proposed blocks.

That is great possibility for the future, @EdgeDLT. We can have the additional speaker as an untrusted node doing such kind of things, which, in my opnion, makes the decentralization of the consensus process more fair.

**Comment by vncoelho** [comment]
Economically speaking, this mempool issue has been here since the free txs.

In the past, we create slots, which solved a lot the problem of flodding the fee txs. 
Then, we created the priorities.

In my opinion, the emergency txs pointed by @Liaojinghui solves the problem because the committee (from my perspective) should be a real-time online system. They should not discuss and take time to solve such situations.
The committee should have standard procedures for that, which would trigger higher minimal networks fees.

In principle, it is GREAT to have the mempool full of txs! There are more fees to be received.
One of the issues is that fees are low at the moment of the attack.

**Comment by Liaojinghui** [comment]
>I believe that stop accepting transactions is critical because we have a logic on the P2P for requesting and receiving them, that would block the tx to be received later, or we would need to not add them on the cache.
Perhaps, create another hash list to request them later.

under the case when the mempool is already full, I think it is ok to halt accepting new transactions. The liveness and stability would definitely be more important than processing new transactions when we already have thousands of unprocessed in the cache. Well, not elegant though.

**Comment by dusmart** [comment]
> In principle, it is GREAT to have the mempool full of txs! There are lots of fees to be collected. One of the issues is that fees are low at the moment of the attack.

I have to reminder that not all of the txs in the mempool can be valid on chain even if they can fit in the mempool now.
The case is that the bad guy can send tons of thousands of transactions to the network whose `ValidUntilBlock` is next block. Then every node's mempool will contain only 50k tx now(the guy should remain at least 50k * GAS_per_tx in his wallet). While only 500 among them can be packed into the next block on mainnet now(5k on testnet). Therefore, only 500 * GAS_per_tx is the real cost for such an experiment.

**Comment by vncoelho** [comment]
I did not mean system fees @dusmart 
Network fees are the ones responsible at this stage.

**Comment by Liaojinghui** [comment]
> I have to reminder that not all of the txs in the mempool can be valid on chain even if they can fit in the mempool now.
The case is that the bad guy can send tons of thousands of transactions to the network whose ValidUntilBlock is next block. Then every node's mempool will contain only 50k tx now(the guy should remain at least 50k * GAS_per_tx in his wallet). While only 500 among them can be packed into the next block on mainnet now(5k on testnet). Therefore, only 500 * GAS_per_tx is the real cost for such an experiment.

A much more reasonable and logical mempool, of course more complex,  strategy then.

**Comment by roman-khimov** [comment]
@Liaojinghui,

> If full, mempool freezes itself and stops accepting new transactions, even if with higher transaction fees.

While this may seem like a fair behavior, I doubt it's viable:
1. What is a "mempool freeze" in a distributed network? Each node has a mempool of its own and can "freeze" with any state weakly related to the state of its neighbor. We can have radically different pools on this network, they will still have this problem.
2. People expect to be able to pay more for priority, they always had this possibility, so taking it out won't make them happy. Imagine a big ICO/mint rush happening, what will people say when they see `OutOfMemory` error from node without any possibility to handle it? They will just say "Neo is broken".
3. Economically CNs don't care about this, they want as much network fees as they can get.

>  no new txs if mempool already full unless 10 times more fee is paid.

However, some economic adjustments can help in avoiding transaction flood. It'll still be possible to cause some mempool difference with carefully planted nodes in different parts of the network, but at least there won't be a lot of traffic on the network, so it'll be easier to deal with it. IIRC now you only need to have bigger `FeePerByte` and given average transaction size of 250 this means paying 0.0000025 GAS more. 10× is just too much IMO, but if mempool eviction is to cost more than it costs now (like you need to pay 10% more) then the cost of the attack will quickly rise to unreasonable levels.

It can be done before reaching mempool capacity as well, have 50% full -> require 10% higher `FeePerByte` than the mempool average, have 70% full -> require 30% higher `FeePerByte`, 90% -> 50% higher. If it's tied to mempool average then even if someone to try pushing more and more, the cost will get higher and higher with each transaction.

> focusing on the mempool would pose a minimal impact on existing neo protocols.

That's true, btw. This economic disincentivization should be easier to start with, maybe it'll even be enough for public networks (but some knobs will be needed for tests/private installations).

@EdgeDLT,

> proposer-builder separation

That's an interesting thing in general, but it's more about the long-standing issue of missing economic motivation for running a Neo node (and some other things like potentially improving MEV situation). The problems here are:
 * potential performance impact both in terms of TPS, but even more in potential additional delays
 * no matter how batches are created all CNs still need to be synchronized about them
 * we better not send these batches with full transaction bodies (especially if anyone can propose), so there is the same problem of where to get a missing transaction from

@vncoelho,

> possibility of emergency transactions

We've got `HighPriorityAttribute` that can be used by committee.

@dusmart,

> While only 500 among them can be packed into the next block on mainnet now(5k on testnet).

Unrelated, but as I've said previously, mainnet should use the same 5K setting as well (it also somewhat mitigates low VUB problem and helps with transaction spikes in general, 500tx per 15s is 33TPS, 5000tx is 333TPS, easy as that).

**Comment by Liaojinghui** [comment]
@roman-khimov I basically agree with our solution and consider yours not much different than mine. when I say freeze the mempool, well just to lazy to consider a proper name, I mean stop processing low fee txs but require higher tx fees, even 10 times more fees. and mempool difference is not a big deal when we already have 50k txs in the cache, users will not see neo as broken, users will see neo as `hot`. But anyway, I agree with our proposal to increase the tx fees step by step.

**Comment by vncoelho** [comment]
> it can be done before reaching mempool capacity as well, have 50% full -> require 10% higher FeePerByte than the mempool average, have 70% full -> require 30% higher FeePerByte, 90% -> 50% higher. If it's tied to mempool average then even if someone to try pushing more and more, the cost will get higher and higher with each transaction.

mempool varies across nodes while acceptance criteria should not. This link to mempool capacity causes other problems.

**Comment by roman-khimov** [comment]
> mempool varies across nodes while acceptance criteria should not. This link to mempool capacity can/may causes other problems.

I've thought about it. Yes, we still can't have perfect synchronization and different parts of the network will have some different sets of transactions potentially with different average `FeePerByte`. But is it a problem? I can run a node already that has a mempool of 10K. You can say my node is broken, but I can run with 100K as well, in any event, my node can have a different acceptance criteria even today. Also, my mempool may already have 50K transactions and I'll reject any new ones unless they pay more while some node may still have some 100 spare and will accept anything. Then all of the nodes are interconnected and constantly communicating with each other, so eventually I'd expect them to have some similar average `FeePerByte`, because if one node has 30% higher average it'll be able to push these pricey transactions to other nodes until their average is about the same.

If won't be perfect, it can't be, but it can work good enough for the purpose of preventing the flood. At least, we can try and this attempt won't cost a lot. Then `RecoverTx` will solve the other part of the problem.

**Comment by cschuchardt88** [comment]
# Design: Mempool Flood Resilience and Consensus Recovery

**Status:** Draft  
**Related:** [neo-project/neo#2862](https://github.com/neo-project/neo/issues/2862), [neo-project/neo#2058](https://github.com/neo-project/neo/issues/2058)  
**Scope:** Mempool admission, P2P prioritization, dBFT recovery under valid-tx flood  
**Non-goals:** Proposer–builder separation, state/snapshot sync, full redesign of dBFT  
**Audience:** Core / consensus / P2P maintainers  

---

## 1. Problem statement

Under a sustained flood of **valid**, fee-paying transactions (reported around **~10k tx/s** in #2862), committee nodes experience:

1. **Mempool churn and divergence** — ascending fee/priority and multi-path injection rotate pool contents so CNs no longer share the same set.
2. **`TxNotFound` on PrepareRequest** — dBFT proposals carry transaction **hashes**; backups that lack bodies spend the view fetching them.
3. **View-change storms** — missed fetches expire into `ChangeView`, often with reason `TxNotFound`.
4. **Consensus-plane starvation** — under load, even high-priority consensus frames (`PrepareRequest`, `ChangeView`, etc.) can fail to propagate in time.

This is a **liveness** issue under adversarial-but-valid load, not classical invalid-spam filtering.

### 1.1 Attack economics (important)

An attacker need not permanently fund a full mempool:

- Fill capacity with txs whose `ValidUntilBlock` is near (e.g. next block).
- Only `MaxTransactionsPerBlock` of them can land on-chain per block; the rest expire.
- Real burn is closer to **fees of txs that actually execute / occupy scarce block space**, not 50k × full fee forever.

Mitigations must raise **churn cost** and protect **consensus progress**, not only “reject invalid junk.”

### 1.2 Design constraints

| Constraint | Implication |
|------------|-------------|
| Mempools will never be perfectly synchronized | Do not design for perfect sync; design for recovery |
| Optimistic dBFT path should stay hash-based | Do not put full block bodies in every PrepareRequest |
| CNs want network fees | Hard “freeze mempool, no fee bidding” fights incentives and UX |
| Public + private networks | Policy knobs must be configurable |

---

## 2. Goals and non-goals

### Goals

1. Keep **optimistic** consensus bandwidth low (hash-only PrepareRequest when mempools are “good enough”).
2. Guarantee a **priority lane** so consensus messages make progress under tx flood.
3. Provide an explicit **missing-transaction recovery** path before / during view change.
4. Raise the **economic cost of mempool churn** via admission/replacement policy.
5. Expose **metrics** so load tests and mainnet ops are measurable.
6. Ship in **incremental PRs** that can land independently where possible.

### Non-goals

- Replacing dBFT.
- Always embedding full tx bodies in PrepareRequest (rejected as default).
- Perfect global mempool identity.
- Solving MEV / PBS in this design.

---

## 3. Current architecture (relevant pieces)

```
[Users / attackers]
        │  tx gossip
        ▼
   RemoteNode / P2P  ──► MemPool (local)
        │
        │  inv / getdata
        ▼
   ConsensusService (dBFT)
        │  PrepareRequest(hashes)
        ▼
   Backups: fetch missing txs → PrepareResponse / ChangeView(TxNotFound)
```

**Happy path today:** primary proposes hashes; backups already have bodies; consensus proceeds.

**Failure path under flood:** bodies missing → fetch storm → timers → CV → possible consensus lag while P2P is saturated with `tx` traffic.

---

## 4. Design overview

Three cooperating layers:

```
┌─────────────────────────────────────────────────────────────┐
│  Layer A — Mempool admission & economics                    │
│  Progressive fee floors, replacement cost, VUB-aware caps   │
└────────────────────────────┬────────────────────────────────┘
                             │ reduces churn rate
┌────────────────────────────▼────────────────────────────────┐
│  Layer B — P2P priority isolation                           │
│  Consensus plane always ahead of tx/inv plane               │
└────────────────────────────┬────────────────────────────────┘
                             │ protects message delivery
┌────────────────────────────▼────────────────────────────────┐
│  Layer C — Consensus recovery                               │
│  Neighbor fetch → RecoverTx → optional CV bitmask           │
│  Happy path: hash-only PrepareRequest unchanged             │
└─────────────────────────────────────────────────────────────┘
```

**Principle:** optimistic path unchanged and cheap; recovery is explicit, high-priority, and incomplete-mempool-aware.

---

## 5. Layer A — Mempool admission and economics

### 5.1 Progressive fee floor (local DoS defense)

Nodes already differ in pool size and contents. Local admission rules may differ; that is acceptable for flood defense.

**Proposal (configurable):**

| Local fill level (`Count / Capacity`) | Min network fee vs reference |
|--------------------------------------|------------------------------|
| `< SoftThreshold` (default 50%) | Existing rules only |
| `≥ SoftThreshold` | `FeePerByte ≥ AvgFeePerByte * (1 + SoftPremium)` (default +10%) |
| `≥ HardThreshold` (default 70%) | `+ HardPremium` (default +30%) |
| `≥ CriticalThreshold` (default 90%) | `+ CriticalPremium` (default +50%) |
| Full | Replacement only if `FeePerByte ≥ VictimFeePerByte * (1 + ReplacePremium)` (default +10% minimum; raise if needed) |

- **Reference average:** exponential moving average of `FeePerByte` over txs currently in the pool (or last N admits), not a global chain parameter.
- **HighPriority / committee attributes:** may bypass soft floors (existing high-priority semantics), not necessarily hard floors—product choice (see Open Questions).

**Rationale:** Attacker who rotates the pool pays escalating premiums; honest users under mild load are unaffected.

### 5.2 VUB-aware capacity

Limit how much of the pool may be occupied by txs that expire very soon:

| Class | Default cap of capacity | Intent |
|-------|-------------------------|--------|
| `ValidUntilBlock ≤ CurrentIndex + 1` | e.g. 25% or `≤ MaxTransactionsPerBlock * K` | Stops “dies next block” spam monopolizing the pool |
| `ValidUntilBlock ≤ CurrentIndex + 2` | e.g. 50% cumulative | Softens short-horizon floods |
| Longer VUB | Remainder | Prefer durable demand |

Exact numbers are knobs; ship defaults, load-test, adjust.

### 5.3 Explicitly reject

- Hard freeze with “no higher fee accepted.”
- Global consensus rules that require identical mempool averages.

### 5.4 Optional capacity / policy (network operators)

- Document raising mainnet `MaxTransactionsPerBlock` toward testnet-like capacity if product goal is higher sustained TPS (Roman’s note in #2862). That is **policy**, orthogonal to recovery protocol, but reduces “only 500 land per block” attack-cost asymmetry.

---

## 6. Layer B — P2P priority isolation

### 6.1 Two planes

| Plane | Examples | Scheduling |
|-------|----------|------------|
| **Consensus** | PrepareRequest/Response, Commit, ChangeView, Recovery*, RecoverTx | Strict priority; reserved bandwidth / queues |
| **Gossip** | tx, inv (tx), getdata(tx), addr | Best-effort; rate limited |

\*Existing recovery messages stay on the consensus plane.

### 6.2 Mechanisms

1. **Egress priority queues** on `RemoteNode` (or shared outbound scheduler): always dequeue consensus before gossip when both pending.
2. **Ingress classification:** parse cheaply; consensus commands never wait behind a long tx verify queue.
3. **Per-peer and global tx ingress caps** (bytes/s and/or tx/s), with headroom reserved for consensus.
4. **Under critical load:** drop or delay new unsolicited `tx` before dropping consensus.

### 6.3 Success criteria

Under a 10k tx/s inject:

- p99 latency for CN→CN consensus messages remains within a fraction of the view timer (target: load-test defined, e.g. &lt; 1s on lab links).
- CV due to pure “message never arrived” falls sharply vs baseline.

---

## 7. Layer C — Consensus recovery (core protocol work)

### 7.1 Happy path (unchanged)

1. Primary builds block from local mempool.
2. PrepareRequest contains **hashes only** (plus existing fields).
3. Backups with full bodies → PrepareResponse as today.

### 7.2 Missing-body timeline (per backup, view *v*)

Let `T_cv` be the ChangeView timer for this view.

```
t=0     PrepareRequest received (or recovered)
        Compute Missing = { hash in proposal | body not in mempool/cache }

t=0+    If Missing empty → PrepareResponse

        Else: request bodies from neighbors (existing getdata / inv paths)
              Mark state: FetchingNeighbors

t=T1    T1 = T_cv / 3   (config: RecoverNeighborFraction, default 1/3)
        If Missing still non-empty → send RecoverTx (high priority)
              Mark state: RecoveringFromPrimary

t=T2    T2 = T_cv * 2/3 or T_cv
        If still missing → ChangeView(TxNotFound) + optional bitmask
```

Heuristics (config, defaults load-testable):

- Skip RecoverTx if `|Missing| / |Proposal| < MissingRatioSkip` (default 0.20) **and** neighbor fetch still in progress until T1.
- Skip RecoverTx if proposal size &lt; `SmallProposalSkip` fraction of MaxTransactionsPerBlock (default 0.20)—small proposals may clear via neighbors alone.
- Always allow RecoverTx if Missing is non-empty at T1 under flood metrics (optional override).

### 7.3 New message: `RecoverTx`

**Direction:** Backup → network (broadcast or CN-directed); **processed by current primary** for view *v* (others may cache or ignore).

**Payload (sketch):**

| Field | Type | Notes |
|-------|------|--------|
| View / block index | as other consensus msgs | Bind to active proposal |
| Proposal fingerprint | hash of PrepareRequest or block hash | Avoid responding to stale views |
| Format | enum | `Empty` \| `MissingBitmask` \| `MissingHashes` |
| Bitmask or hash list | variable | Prefer bitmask when ordered proposal hashes known |
| Signature | CN sig | Authenticate sender as committee |

**Empty / parameter-less variant:** “I am missing something for this proposal” — primary may re-broadcast full missing set if it tracks per-CN state; bitmask is preferred for bandwidth.

**Priority:** consensus plane (Layer B).

### 7.4 Primary behavior on RecoverTx

1. Validate signer is committee, view matches, proposal matches.
2. Collect RecoverTx until:
   - **`f+1` distinct CNs** report missing data for this proposal, **or**
   - timer `RecoverAggregateTimeout` (default remaining until ~2/3 of CV).
3. Compute **UnionMissing** across reports (bitmask OR / hash set union).
4. Emit **`RecoverTxResponse`** (or reuse a dedicated consensus inventory message):

| Mode | Content |
|------|---------|
| **Delta** (preferred) | Bodies only for UnionMissing |
| **Full** (fallback) | All proposal bodies if delta construction fails or set is huge |

5. Backups insert bodies into a **proposal-scoped cache** (not necessarily general mempool eviction victims) and continue verification → PrepareResponse.

**Idempotency:** Duplicate RecoverTx from same CN ignored; response may be re-broadcast; receivers ignore duplicate bodies.

### 7.5 ChangeView enrichment (ties to #2058)

When sending CV for `TxNotFound`:

- Include **reason code** (existing or extended).
- Include **bitmask / compact missing set** relative to last PrepareRequest hashes.

Next primary (or same after view bump) uses:

- Local mempool, plus
- Aggregated missing bitmasks from CVs,

to prefer a proposal that maximizes expected readiness (e.g. intersection of widely available txs)—exact algorithm can start simple: **prefer txs present locally that appear in ≥ f+1 readiness hints**, fill remainder by fee.

This does not require full #2058; it is compatible with it.

### 7.6 Explicitly not the default

**Full transaction bodies in every PrepareRequest**

- Allowed only as **emergency escape hatch** (config `PrepareRequestIncludeBodies=false` by default):
  - After `N` consecutive views failing with TxNotFound, or
  - After RecoverTx path fails for the same proposal hash.
- Cost: multi-MB messages, multi-second propagation on modest links—acceptable only as last resort.

---

## 8. State machines (backup)

```
                 PrepareRequest
                      │
                      ▼
              ┌───────────────┐
              │  HaveBodies?  │──yes──► PrepareResponse
              └───────┬───────┘
                      │ no
                      ▼
              FetchNeighbors ──success──► PrepareResponse
                      │
                   timeout T1
                      ▼
                 RecoverTx sent
                      │
              ┌───────┴────────┐
           bodies            timeout T2
              │                  │
              ▼                  ▼
        PrepareResponse    ChangeView(+bitmask)
```

Primary adds: **AggregateRecoverTx → RecoverTxResponse**.

---

## 9. Configuration (illustrative)

| Name | Default | Layer |
|------|---------|--------|
| `MempoolSoftFillThreshold` | 0.50 | A |
| `MempoolSoftFeePremium` | 0.10 | A |
| `MempoolHardFillThreshold` | 0.70 | A |
| `MempoolHardFeePremium` | 0.30 | A |
| `MempoolCriticalFillThreshold` | 0.90 | A |
| `MempoolCriticalFeePremium` | 0.50 | A |
| `MempoolReplacePremium` | 0.10 | A |
| `MempoolShortVubShare` | 0.25 | A |
| `TxIngressBytesPerSecond` | lab-tuned | B |
| `ConsensusEgressAbsolutePriority` | true | B |
| `RecoverNeighborFraction` | 0.33 | C |
| `RecoverAggregateTimeout` | view-relative | C |
| `MissingRatioSkip` | 0.20 | C |
| `PrepareRequestIncludeBodies` | false | C |
| `PrepareRequestBodiesAfterFailedViews` | 3 | C |

Private nets may disable A/B throttles for testing.

---

## 10. Metrics

| Metric | Use |
|--------|-----|
| `consensus_tx_not_found_total` | Primary symptom |
| `consensus_recover_tx_sent/received` | Recovery usage |
| `consensus_recover_tx_filled` | Success of recovery |
| `consensus_view_change_reason{reason}` | Histogram of CV causes |
| `mempool_admit_reject{reason}` | fee_floor, vub_cap, full |
| `mempool_avg_fee_per_byte` | Floor reference |
| `p2p_egress_queue_delay{plane}` | consensus vs gossip |
| `p2p_tx_ingress_dropped` | Throttle effectiveness |
| `consensus_message_e2e_latency` | Priority isolation proof |

---

## 11. Testing strategy

### 11.1 Unit / component

- Mempool: fee floors at thresholds; VUB caps; replacement premiums.
- Priority scheduler: consensus never starved behind N pending txs.
- RecoverTx: aggregation at f+1; stale view ignored; delta vs full response.
- CV bitmask encode/decode.

### 11.2 Integration / load (required for merge of Layer C)

Scenario matrix:

| Scenario | Inject | Expect |
|----------|--------|--------|
| Baseline | low load | no RecoverTx; normal blocks |
| Medium flood | ~5k tx/s | fees rise; consensus holds |
| Hard flood | ~10k tx/s multi-path | RecoverTx engages; CV rate bounded |
| Partitioned inject | different tx sets per CN | recovery or informed CV, not permanent stall |
| Disable Layer A | flood | Layer B+C still protect liveness (degraded) |

Success criteria examples (lab-defined absolute numbers later):

- Median block time increase &lt; X% under hard flood vs baseline.
- No multi-minute consensus halt in 30-minute soak.
- RecoverTx bandwidth ≪ full PrepareRequest-with-bodies every view.

### 11.3 Cross-implementation

Any new consensus message must be specified for **neo-go** compatibility (wire format, priority, state machine). Prefer a short NEP/RNP-style wire appendix before mainnet.

---

## 12. Alternatives considered

| Alternative | Decision | Why |
|-------------|----------|-----|
| Always pack bodies in PrepareRequest | **Reject as default** | Large fixed cost; hurts healthy network; multi-second hops on modest links |
| Freeze mempool when full | **Reject** | UX (“broken under load”); local freezes diverge; kills fee market |
| Ingress throttle only | **Insufficient alone** | Many senders bypass; doesn’t fix existing divergence |
| PBS / external block building | **Out of scope** | Doesn’t remove body-sync problem; adds latency |
| Rely only on longer CV timers | **Insufficient** | Masks symptoms; doesn’t restore bodies or priority |
| Progressive fees only | **Phase 1 only** | Cheap; may be enough for mild public attacks; not sufficient alone for adversarial multi-path |

---

## 13. Security and privacy notes

- RecoverTx must be **committee-authenticated** to avoid public amplification.
- Bitmasks leak which txs a CN lacks—acceptable among CNs; do not require public broadcast of full missing lists if avoidable.
- Primary must cap RecoverTxResponse size (chunking if needed) to avoid self-DoS.
- Proposal-scoped body cache should bound memory (drop on view change / height advance).

---

## 14. Rollout

1. **Phase 1 (no wire change):** Layer A + B + metrics. Deploy to testnet; measure CV / TxNotFound under synthetic flood.
2. **Phase 2 (wire change):** RecoverTx + response + optional CV bitmask. Dual-impl agreement; testnet only until soak passes.
3. **Phase 3:** Escape-hatch PrepareRequest bodies (off by default); tune defaults from Phase 1–2 data.
4. **Mainnet:** Policy review (fee floors, MaxTransactionsPerBlock) separate from recovery enablement.

---

## 15. Key Decisions

| Decision | Choice | Rationale |
|----------|--------|-----------|
| Mempool consistency | Accept inconsistency | Unavoidable under multi-sender flood |
| PrepareRequest content | Hashes by default | Preserve dBFT optimistic bandwidth |
| Full bodies in proposal | Last-resort only | Avoid multi-second consensus messages every block |
| Hard mempool freeze | No | Bad UX and incentives |
| Economic defense | Progressive local fee floors + replace premium | Raises churn cost; minimal protocol change |
| VUB spam | Cap short-VUB share of pool | Closes cheap “expire next block” rotation |
| Consensus under load | Strict P2P priority isolation | Fixes observed ChangeView delivery failure |
| Recovery | RecoverTx after neighbor fetch, primary aggregates f+1 | Fills gaps without always shipping full blocks |
| CV | Enrich with TxNotFound + bitmask | Improves next proposal / aligns with #2058 direction |
| Implementation home | MemPool + P2P + DBFTPlugin | Not LocalNode connection management |

---

## 16. Open Questions

1. **HighPriority txs vs fee floors** — Full bypass of soft/hard floors, or only soft?
2. **RecoverTx topology** — True broadcast vs direct send to primary only (direct saves bandwidth; broadcast helps if primary identity lag)?
3. **Bitmask size** — Cap proposal length for bitmask practicality; fallback to hash list if proposal huge?
4. **Mainnet MaxTransactionsPerBlock** — Raise toward 5k as part of this work or separate governance?
5. **Protocol version / hardfork** — New consensus messages need capability bit or always-on after height?

*(Answers can be filled without blocking Phase 1.)*

---

## 17. PR Plan

### PR1 — Metrics and load-test harness hooks
- **Title:** `observability: consensus TxNotFound and mempool admit metrics`
- **Affects:** DBFTPlugin, MemPool, optional P2P counters
- **Depends on:** none
- **Description:** Counters/histograms listed in §10; baseline dashboards for lab floods. No behavior change.

### PR2 — Mempool progressive fees and VUB caps
- **Title:** `mempool: progressive fee floors and short-VUB capacity limits`
- **Affects:** `MemoryPool` / admission paths, config, unit tests
- **Depends on:** PR1 (optional but preferred)
- **Description:** Layer A. Config defaults conservative; document operator knobs.

### PR3 — P2P consensus priority and tx ingress limits
- **Title:** `p2p: prioritize consensus plane and rate-limit tx ingress`
- **Affects:** RemoteNode / outbound queues / message handling
- **Depends on:** PR1
- **Description:** Layer B. Ensure consensus egress/ingress cannot be starved by tx flood.

### PR4 — RecoverTx protocol (wire + primary aggregate)
- **Title:** `dbft: RecoverTx missing-body recovery for PrepareRequest`
- **Affects:** DBFTPlugin messages, state machine, proposal body cache
- **Depends on:** PR3 (priority), ideally PR1
- **Description:** Layer C core. Spec + C# impl; coordinate neo-go tracking issue.

### PR5 — ChangeView TxNotFound bitmask / readiness hints
- **Title:** `dbft: attach missing-tx bitmask to TxNotFound view changes`
- **Affects:** CV payload, next-primary proposal heuristics
- **Depends on:** PR4 (shared proposal indexing)
- **Description:** Improves subsequent views under persistent divergence.

### PR6 — Escape hatch: optional bodies in PrepareRequest
- **Title:** `dbft: optional PrepareRequest transaction bodies after repeated failures`
- **Affects:** PrepareRequest serialization, config
- **Depends on:** PR4–5
- **Description:** Disabled by default; enable after N failed views or explicit config.

### PR7 — Load-test suite and default tuning
- **Title:** `test: mempool flood consensus soak and default parameter tuning`
- **Affects:** tests / tools / docs
- **Depends on:** PR2–5
- **Description:** Reproducible 5k/10k inject scenarios; lock default knobs from data.

---

## 18. Summary

Issue #2862 is a real **liveness under valid flood** problem at the intersection of mempool economics, P2P scheduling, and hash-based dBFT proposals. The design:

1. **Slows churn** (fees + VUB caps),  
2. **Protects the consensus plane** (priority isolation),  
3. **Recovers missing bodies deliberately** (RecoverTx),  
4. **Keeps the optimistic path cheap** (no full bodies in every PrepareRequest).

Ship **A + B first** (no wire break), then **C** with dual-impl agreement, then optional escape hatches and tuning.
